go/libraries/doltcore/branch_control/expr_parser_node.go

```diff
@@ -0,0 +1,372 @@
+				// We have more sort orders on the root, but no more in our expressions, so we put the remaining root
+				// sort orders as a divergence and set this as a destination node
+				root.Divergences = []MatchNode{{
+					SortOrders:  remainingRootSortOrders[1:],
+					Divergences: root.Divergences,
+					Data:        root.Data,
+				}}
+				root.SortOrders = root.SortOrders[:len(root.SortOrders)-len(remainingRootSortOrders)+1]
+				root.Data = &data
+				break
+			} else if len(remainingRootSortOrders) == 1 && i < allSortOrdersMaxIndex {
+				// We've run out of sort orders on the root, but still have more from the expressions, so check if a
+				// divergence will match the next sort order from the expressions
+				nextSortOrder := allSortOrders[i+1]
+				for divergenceIdx, divergence := range root.Divergences {
+					if divergence.SortOrders[0] == nextSortOrder {
+						remainingRootSortOrders = divergence.SortOrders
+						root = &root.Divergences[divergenceIdx]
+						continue ParentLoop
+					}
+				}
+				// None of the divergences matched, so we create a new one and add it. As we're using a pool, we need to
+				// create a new slice.
+				originalSortOrders := allSortOrders[i+1:]
+				newSortOrders := make([]int32, len(originalSortOrders))
+				copy(newSortOrders, originalSortOrders)
+				root.Divergences = append(root.Divergences, MatchNode{
+					SortOrders:  newSortOrders,
+					Divergences: nil,
+					Data:        &data,
+				})
+				break
+			} else {
+				// We have no more sort orders on either side so this is an exact match, therefore we update the data
+				root.Data = &data
+				break
+			}
+		} else {
+			// Since the sort orders do not match, we create a divergence here with the remaining expressions' sort
+			// orders, and move the root's remaining sort orders to its own divergence.
+			splitRoot := MatchNode{
+				SortOrders:  remainingRootSortOrders,
+				Divergences: root.Divergences,
+				Data:        root.Data,
+			}
+			// As we're using a pool, we need to create a new slice
+			originalSortOrders := allSortOrders[i:]
+			newSortOrders := make([]int32, len(originalSortOrders))
+			copy(newSortOrders, originalSortOrders)
+			newDivergence := MatchNode{
+				SortOrders:  newSortOrders,
+				Divergences: nil,
+				Data:        &data,
+			}
+			root.SortOrders = root.SortOrders[:len(root.SortOrders)-len(remainingRootSortOrders)]
+			root.Divergences = []MatchNode{splitRoot, newDivergence}
+			// As the root's data is now in the split, we set the data here to nil as it's no longer a destination node.
+			// If it wasn't a destination node, then nothing changes (we just set the split's data to nil as well).
+			root.Data = nil
+			break
+		}
+	}
+}
+
+// Remove will remove the given expressions to the node hierarchy. If the expressions do not exist, then nothing
+// happens. Assumes that the given expressions have already been folded.
+func (mn *MatchNode) Remove(databaseExpr, branchExpr, userExpr, hostExpr string) uint32 {
```

Comment:
Without reading this closely, I'll just note that it's odd that there appears to be so little shared tree walking code. I would expect something like FindParent(entry) that returns the parent node a leaf should be added to, that you could then use for both Insert and  Delete operations. Odd that it's not the case.

Reply:
There's not really a lot of code that would meaningfully be shared between the two. The part that leads up to finding a divergence is the same, but the context of how and why a match wasn't found isn't shared between the two, and needing to return all of that information from a common search function would just result in more code than what's already here.